### PR TITLE
Concurrent hash `LEFT` and `INNER` join

### PR DESCRIPTION
**Pull request by lgbo-ustc**
now adding joined blocks from right table can be run parallelly, speedup the join process

### Changelog category (leave one):
- Performance improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

The default `HashJoin` is not thread safe for inserting right table's rows and run it in a single thread. When the right table is large, the join process is too slow with low cpu utilization.

We create multiple `HashJoin` instances here. In `addJoinedBlock()`, one input block is split into multiple blocks corresponding to the `HashJoin` instances by hashing every row on the join keys. And make a guarantee that every `HashJoin` instance is written by only one thread at the same time.

When come to the left table matching, the blocks from left table are also split into different `HashJoin` instances.

We take tests on a table with 19 million rows with following query
```SQL
SELECT
    t1.type,
    count(*)
FROM
(
    SELECT
        ip,
        rtime,
        id,
        type
    FROM test_table
    WHERE (day = '2022-02-01') AND (hour = '00')
) AS t1
ANY INNER JOIN
(
    SELECT
        ip,
        rtime,
        id,
        type
    FROM test_table
    WHERE (day = '2022-02-01') AND (hour = '00')
) AS t2 ON (t1.ip = t2.ip) AND (t1.id = t2.id) AND (t1.rtime = t2.rtime) AND (t1.type = t2.type)
GROUP BY t1.type
``` 

As a baseline, we use `HashJoin` by limit both left table and right table are  handled in a single thread.
| threads   |      elapsed time(s)   |
|----------|:-------------:|
| 1|  35.35 |

and then run the `ConcurrentHashJoin`, we have the following result
| threads   |      elapsed time(s)   |
|----------|:-------------:|
| 1|  34.57 |
| 2|  22.22 |
| 4|  12.28 |
| 8|  7.62 |
| 16|  5.9 |

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+            {
+                mutable_final_columns[pos]->insertRangeFrom(*from_column.column, 0, from_column.column->size());
+            }
+        }
+    }
+
+    size_t i = 0;
+    for (auto & name_and_type : names_and_types)
+    {
+        ColumnPtr col_ptr = std::move(mutable_final_columns[i]);
```
Comment (zhanglistar) [inline src/Interpreters/ConcurrentHashJoin.cpp:126]: merge

File: src/Core/Settings.h [hunk 2]
```diff
@@ -351,6 +351,7 @@ class IColumn;
     M(UInt64, partial_merge_join_left_table_buffer_bytes, 0, "If not 0 group left table blocks in bigger ones for left-side table in partial merge join. It uses up to 2x of specified memory per joining thread.", 0) \
     M(UInt64, partial_merge_join_rows_in_right_blocks, 65536, "Split right-hand joining data in blocks of specified size. It's a portion of data indexed by min-max values and possibly unloaded on disk.", 0) \
     M(UInt64, join_on_disk_max_files_to_merge, 64, "For MergeJoin on disk set how much files it's allowed to sort simultaneously. Then this value bigger then more memory used and then less disk I/O needed. Minimum is 2.", 0) \
+    M(Bool, enable_parallel_hash_join, false, "Enable paralle join algorithm.", 0)\
```
Comment (Algunenano) [inline src/Core/Settings.h:354]: Basic question: What are the reasons for this to be disabled by default? Seems like a great addition to have it always on.

Also, there is a typo: `paralle` -> `parallel`

Comment (lgbo-ustc) [reply src/Core/Settings.h:354]: It maybe have unexpected behaviors, we want to keep it in experiement status

File: src/Interpreters/ConcurrentHashJoin.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    bool supportParallelJoin() const override { return true; }
+    std::shared_ptr<NotJoinedBlocks>
+    getNonJoinedBlocks(const Block & left_sample_block, const Block & result_sample_block, UInt64 max_block_size) const override;
+private:
+    struct InnerHashJoin
+    {
+        std::mutex mutex;
+        std::unique_ptr<HashJoin> data;
+        std::list<Block> pending_right_blocks;
+        std::atomic<bool> in_inserting = false;
```
[the remark this answers is not in the document]
Comment (lgbo-ustc) [reply src/Interpreters/ConcurrentHashJoin.h:57]: `rows` maybe updated by threads at the same time, must be atomic

File: src/Core/Settings.h (hunk 2 above)
Comment (lgbo-ustc) [reply src/Core/Settings.h]: done

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    if (table_join->strictness() == ASTTableJoin::Strictness::Asof ||
+        table_join->strictness() == ASTTableJoin::Strictness::Semi ||
+        !isRightOrFull(table_join->kind()))
+    {
+        return {};
+    }
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid join type. join kind: {}, strictness: {}", table_join->kind(), table_join->strictness());
+}
+
+std::shared_ptr<ExpressionActions> ConcurrentHashJoin::buildHashExpressionAction(const Block & block, const Strings & based_columns_names, Strings & hash_columns_names)
```
Comment (vdimir) [inline src/Interpreters/ConcurrentHashJoin.cpp:194]: Why do we need such complex way to calculate hash? We can just access data in Block and calculate what we need

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+        auto & hash_join = hash_joins[i];
+        auto & dispatched_block = dispatched_blocks[i];
+        auto rows = dispatched_block.rows();
+        check_total_rows += rows;
+        check_total_bytes += dispatched_block.bytes();
+        std::unique_lock lock(hash_join->mutex);
+        
+        // Don't take the real insertion here, because inserting a block into HashTable is a time-consuming operation,
+        // it may cause serious lock contention and make the whole process slow.
+        hash_join->pending_right_blocks.emplace_back(std::move(dispatched_block));
```
Comment (vdimir) [inline src/Interpreters/ConcurrentHashJoin.cpp:69]: Did you test implementation without this queue? it seems a bit over-complicated. Here we have separate mutex for each `hash_join`, so calls `ConcurrentHashJoin::addJoinedBlock` would block each other only when accessing same `hash_join` instance. Maybe make one iteration with `try_lock` to insert into available slots and then another loop with waiting.

File: tests/queries/1_stateful/00172_parallel_join.sql [hunk 6]
```diff
@@ -0,0 +1,204 @@
+set join_algorithm='parallel_hash';
```
Comment (vdimir) [inline tests/queries/1_stateful/00172_parallel_join.sql:1]: Did you compare results with regular algorithm?

**Review by vdimir — COMMENTED** [review]
Thank you for such big effort!

Couple of concerns: first is way of calculating hash via building and parsing query, and second is logic with pending blocks seems a bit over-complicated, maybe we can simplify it somehow.

File: tests/queries/0_stateless/01911_logical_error_minus.sql [hunk 7]
```diff
@@ -26,7 +26,7 @@ INSERT INTO codecTest (key, name, ref_valueF64, valueF64, ref_valueF32, valueF32
 
 SELECT IF(-2, NULL, 0.00009999999747378752), IF(1048577, 1048576, NULL), c1.key, IF(1, NULL, NULL), c2.key FROM codecTest AS c1 , codecTest AS c2 WHERE ignore(IF(257, -2, NULL), arrayJoin([65537]), IF(3, 1024, 9223372036854775807)) AND IF(NULL, 256, NULL) AND (IF(NULL, '1048576', NULL) = (c1.key - NULL)) LIMIT 65535;
 
-SELECT c1.key, c1.name, c1.ref_valueF64, c1.valueF64, c1.ref_valueF64 - c1.valueF64 AS dF64, '', c2.key, c2.ref_valueF64 FROM codecTest AS c1 , codecTest AS c2 WHERE (dF64 != 3) AND c1.valueF64 AND (c2.key = (c1.key - 1048576)) LIMIT 0;
```
[the remark this answers is not in the document]
Comment (lgbo-ustc) [reply tests/queries/0_stateless/01911_logical_error_minus.sql:29]: using c1.valueF64 as a filter column throws a invalid type exception

File: src/Interpreters/ConcurrentHashJoin.cpp (hunk 4 above)
Comment (lgbo-ustc) [reply src/Interpreters/ConcurrentHashJoin.cpp:194]: Do you have any example to calculate hash from block data directly ?
This is simplest the way that I have found

File: src/Interpreters/ConcurrentHashJoin.cpp (hunk 5 above)
Comment (lgbo-ustc) [reply src/Interpreters/ConcurrentHashJoin.cpp:69]: Maybe it takes more time, but We could have a test on this

File: src/QueryPipeline/QueryPipelineBuilder.cpp [hunk 8]
```diff
@@ -336,15 +338,39 @@ std::unique_ptr<QueryPipelineBuilder> QueryPipelineBuilder::joinPipelines(
...
+        right->resize(num_streams);
+        auto concurrent_right_filling_transform = [&](OutputPortRawPtrs outports)
+        {
+            Processors processors;
+            for (auto & outport : outports)
+            {
+                auto adding_joined = std::make_shared<FillingRightJoinSideTransform>(right->getHeader(), join);
+                connect(*outport, adding_joined->getInputs().front());
+                processors.emplace_back(adding_joined);
+            }
```
[the remark this answers is not in the document]
Comment (lgbo-ustc) [reply src/QueryPipeline/QueryPipelineBuilder.cpp:357]: `resize(1)` makes a finished signal for the whole up stream here.

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+        std::move(actions),
+        prepared_sets,
+        subqueries_for_sets,
+        true, false, true, false);
+    ActionsVisitor(visitor_data, visit_log.stream()).visit(func_ast);
+    actions = visitor_data.getActions();
+    return std::make_shared<ExpressionActions>(actions);
+}
+
+ConcurrentHashJoin::BlockDispatchControlData & ConcurrentHashJoin::getBlockDispatchControlData(const Block & block, TableIndex table_index)
```
[the remark this answers is not in the document]
Comment (lgbo-ustc) [reply src/Interpreters/ConcurrentHashJoin.cpp]: have modified the interfaces in `ExpressionAnalyzer.cpp` to pass left table header in.

File: src/Interpreters/ConcurrentHashJoin.cpp (hunk 4 above)
Comment (lgbo-ustc) [reply src/Interpreters/ConcurrentHashJoin.cpp]: Found that , `Blocks DistributedSink::splitBlock(const Block & block)` use a simillar way by using expression action

**Comment by lgbo-ustc** [comment]
the test case `01049_join_low_card_bug_long` failed for the reason that the column ordered by has duplicated values.
when we resize the left table's streams, the result's order become undeterministic。

should we update the sql by adding more columns in the order by clause?

**Comment by vdimir** [comment]
> should we update the sql by adding more columns in the order by clause

yes, it's better to make queries in tests deterministic. You also and add new algorithm to this tests

```diff
- {% for join_algorithm in [None, 'partial_merge'] -%}
+ {% for join_algorithm in [None, 'partial_merge', 'parallel_hash'] -%}
```

**Comment by lgbo-ustc** [comment]
I have report a bug issue #36599 for the backward compatibility check failure.

**Comment by lgbo-ustc** [comment]
resize left pipeline would cause some unexpected behaviors when `optimize_read_in_order` is true

**Comment by lgbo-ustc** [comment]
Move the resize actions for left and right pipelines into 'InterpreterSelectQuery' and disable it when `optimize_read_in_order` is true

**Comment by lgbo-ustc** [comment]
I run `02122_join_group_by_timeout`  at local with the same settings, but the result is the same.

**Comment by lgbo-ustc** [comment]
@vdimir  the remaing failures look irrelevant to this pr

File: src/QueryPipeline/QueryPipelineBuilder.cpp [hunk 10]
```diff
@@ -344,15 +347,45 @@ std::unique_ptr<QueryPipelineBuilder> QueryPipelineBuilder::joinPipelines(
...
-        totals_port = adding_joined->addTotalsPort();
+    if (join->supportParallelJoin())
+    {
+        if (!keep_left_read_in_order)
+        {
+            left->resize(max_streams);
+            num_streams = max_streams;
+        }
+
+        if (!right->hasTotals())
```
Comment (vdimir) [inline src/QueryPipeline/QueryPipelineBuilder.cpp:360]: Why not `join->supportParallelJoin() && !right->hasTotals()` ?

File: src/Interpreters/ConcurrentHashJoin.cpp (hunk 4 above)
Comment (vdimir) [reply src/Interpreters/ConcurrentHashJoin.cpp]: @lgbo-ustc See changes in https://github.com/ClickHouse/ClickHouse/tree/concurrent_join (Also you may allow to push this changes directly to PR [link](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)). We don't need create actions, just calculate hash. Also cityHash is to heavy, some weak faster hash would be fine.

File: src/QueryPipeline/QueryPipelineBuilder.cpp (hunk 10 above)
Comment (lgbo-ustc) [reply src/QueryPipeline/QueryPipelineBuilder.cpp]: OK

File: src/Interpreters/ConcurrentHashJoin.cpp (hunk 4 above)
Comment (lgbo-ustc) [reply src/Interpreters/ConcurrentHashJoin.cpp]: updated

**Review by vdimir — APPROVED** [review]
Thank you for your contribution!

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+    ColumnRawPtrs key_cols;
+    for (const auto & key_name : key_columns_names)
+    {
+        key_cols.push_back(from_block.getByName(key_name).column.get());
+    }
+    IColumn::Selector selector(num_rows);
+    for (size_t i = 0; i < num_rows; ++i)
+    {
+        SipHash hash;
```
Comment (alexey-milovidov) [inline src/Interpreters/ConcurrentHashJoin.cpp:180]: SipHash is too slow for this purpose.

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        for (const auto & key_col : key_cols)
+        {
+            key_col->updateHashWithValue(i, hash);
+        }
+        selector[i] = hash.get64() % num_shards;
```
Comment (alexey-milovidov) [inline src/Interpreters/ConcurrentHashJoin.cpp:185]: Division by runtime value is also slow.

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+                {
+                    hash_join->mutex.unlock();
+                    return false;
+                }
+
+                hash_join->mutex.unlock();
+                iter = pending_blocks.erase(iter);
+            }
+            else
+                iter++;
```
Comment (alexey-milovidov) [inline src/Interpreters/ConcurrentHashJoin.cpp:75]: ++iter;

File: src/Core/Settings.h [hunk 14]
```diff
@@ -346,7 +346,7 @@ class IColumn;
     M(UInt64, max_bytes_in_join, 0, "Maximum size of the hash table for JOIN (in number of bytes in memory).", 0) \
     M(OverflowMode, join_overflow_mode, OverflowMode::THROW, "What to do when the limit is exceeded.", 0) \
     M(Bool, join_any_take_last_row, false, "When disabled (default) ANY JOIN will take the first found row for a key. When enabled, it will take the last row seen if there are multiple rows for the same key.", IMPORTANT) \
-    M(JoinAlgorithm, join_algorithm, JoinAlgorithm::HASH, "Specify join algorithm: 'auto', 'hash', 'partial_merge', 'prefer_partial_merge'. 'auto' tries to change HashJoin to MergeJoin on the fly to avoid out of memory.", 0) \
+    M(JoinAlgorithm, join_algorithm, JoinAlgorithm::HASH, "Specify join algorithm: 'auto', 'hash', 'partial_merge', 'prefer_partial_merge', 'parallel_hash'. 'auto' tries to change HashJoin to MergeJoin on the fly to avoid out of memory.", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:349]: Why don't set it by default?

**Comment by alexey-milovidov** [comment]
@vdimir let's add it to settings randomization in tests?

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+namespace JoinStuff
+{
+ConcurrentHashJoin::ConcurrentHashJoin(ContextPtr context_, std::shared_ptr<TableJoin> table_join_, size_t slots_, const Block & right_sample_block, bool any_take_last_row_)
+    : context(context_)
+    , table_join(table_join_)
+    , slots(slots_)
+{
+    if (!slots_ || slots_ >= 256)
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Invalid argument slot : {}", slots_);
```
Comment (alexey-milovidov) [inline src/Interpreters/ConcurrentHashJoin.cpp:37]: Wrong punctuation, need to remove whitespace before colon.

Comment (alexey-milovidov) [reply src/Interpreters/ConcurrentHashJoin.cpp:37]: Why cannot have more than 256?

**Comment by alexey-milovidov** [comment]
Let's replace SipHash with WeakHash32.

**Comment by vdimir** [comment]
@alexey-milovidov 

> Let's replace SipHash with WeakHash32.

It's already done in this PR https://github.com/ClickHouse/ClickHouse/pull/36970/files#diff-209d0f5d4e2c1c1357882988d3093960e44c24afdc77a046deb25e0ee5d87336R191-R195

> Why don't set it by default?

A bit afraid that it can slow down some queries because of splitting and dispatching blocks

> let's add it to settings randomization in tests?

Setting throw force to use selected join algorithm and will throw an error if kind is unsupported (right, full), so it can't be added as is right now. But at least in stress test we can enable it https://github.com/ClickHouse/ClickHouse/pull/37436

**Comment by amosbird** [comment]
This PR is marked as "Performance improvement" but there is no related perf tests. I wonder if the conclusion still holds after https://github.com/ClickHouse/ClickHouse/pull/38052

**Comment by yuanLier** [comment]
Hello, may I ask why this optimization does not support right join and full join? Is it not supported at the principle level or at the implementation level?

**Pull request merged**